### DISCUSSION TITLE
Muse  Qt6 help

### DISCUSSION THREAD

**Discussion by edpil02** [muse-sequencer/muse#1251] [Q&A]
I made a patch to port muse to qt6.
The port seems to work but I have a problem when booting up  and I can't solve it.
I found the faulty code (songfile.cpp) but don't understand why  the error occurs.

![Screenshot_20231103_131243](https://github.com/muse-sequencer/muse/assets/43163048/b6737c26-3f02-48ca-99a3-440091ff3d63)

**Comment by edpil02** [comment]
Perhaps the xml reader ?

**Comment by terminator356** [comment]
Hi, welcome back.
The first two lines of file "default.med" should show this - is it correct over there?
```
<?xml version="1.0"?>
<muse version="3.4">

```

In songfile.cpp, function MusE::read() at line 1174, it is supposed to ignore the first file line 
`<?xml version="1.0"?>`
and proceed to read the next line
`<muse version="3.4">`
where it extracts the "3.4" and compares with our current file version (3.4), and warns if a mismatch exists.

I stepped through the code and it appears to be functioning correctly here.
Can I assume you are comfortable running a debugger and setting breakpoints and checking variables?

Set a breakpoint at line 1180 where it says
`      for (;;) {
`

Run the debugger and when it stops at the breakpoint, step through the code and
 observe carefully how it handles each tag and attribute that it finds.

In particular, watch these two lines for the values of 'token' and 'tag':
```
            MusECore::Xml::Token token = xml.parse();
            const QString& tag = xml.s1();

```
The sequence should progress as follows:

For the first file line, 'token' should say 'Proc', 'tag' should say "xml version=1.0", and it 
should simply fall through and ignore the line, and proceed to the next for loop iteration.

Next, for the second line, 'token' should say "TagStart" and 'tag' should say "muse". 
Nothing special should happen here and it should then move on to the next for loop iteration.

On the next for loop iteration, 'token' should say "Attribut" and 'tag' should say "version".
In this case, the following section at line 1251 should execute:
```
                  case MusECore::Xml::Attribut:
                        if (tag == "version") {
                              int major = xml.s2().section('.', 0, 0).toInt();
                              int minor = xml.s2().section('.', 1, 1).toInt();
                              xml.setVersion(major, minor);
                              }
                        break;

``` 

If this section runs, what are the values of the two variables 'int major' and 'int minor' ?
The should be 3 and 4 respectively.

If you cannot run a debugger, instead try inserting a diagnostic print line at line 1185, like this:
```
            MusECore::Xml::Token token = xml.parse();
            const QString& tag = xml.s1();
            fprintf(stderr, "MusE::read(): token:%d tag:%s\n", token, tag.toLatin1().constData());   // Diagnostics
            switch (token) {

```
Run the app and observe the printout.

Maybe the reader is processing file line 1 when in fact it should be ignoring it?

Good luck.
Tim.

**Comment by terminator356** [comment]
If you cannot run a debugger, then in addition to the diagnostic print line mentioned above, 
 you may want to also add an additional diagnostic print line in the section at line 1251, like this:

```
                  case MusECore::Xml::Attribut:
                        if (tag == "version") {
                              int major = xml.s2().section('.', 0, 0).toInt();
                              int minor = xml.s2().section('.', 1, 1).toInt();
                              fprintf(stderr, "MusE::read(): major:%d minor:%d\n", major, minor);   // Diagnostics
                              xml.setVersion(major, minor);
                              }
                        break;

```

**Comment by edpil02** [comment]
Thanks for your help.

Normally  the  Warning Messagebox prints    " File version is  -1.-1"  not  1.0 
I made some changes and tests  in  the file   muse-4.2.1/libs/xml/xml.cpp  because i think  a parsing error :

Xml::Xml(FILE* _f)
      {
      f          = _f;
      _destStr   = nullptr;
      _destIODev = nullptr;
      _line      = 0;
      _col       = 0;
      level      = 0;
      inTag      = false;
      inComment  = false;
      lbuffer[0] = 0;
      bufptr     = lbuffer;
      _minorVersion = -1;              change to  ->  1 
      _majorVersion = -1;               change to ->   0
      }

What is strange is that the port to qt6 does not affect any of the files concerned. After the bug reading "default.med" the program seems to work normally. I think of subtle QT change ? .Looks like a problem reading  default.med xml file at startup.

I'm going to  try your proposals   then download and  build edb-debugger to  do some tracing.
And I wanted to clarify; I don't know much about programming let alone QT but just a beginner in coding..

Sorry for my english.

**Comment by terminator356** [comment]
OK Thanks for the update. 

If you would like, I can recommend either QtCreator or KDevelop. 
Both are excellent Integrated Development Environments (IDE).
Each one makes coding and debugging together a better experience.
Both are usually found in major linux distribution package managers. Easy to install a package.

Of course, there may be some extra learning involved.
But it is usually worth the time investment.

No worries, which ever way works best for you.

Tim.

**Comment by edpil02** [comment]
After giving up once, I resumed debugging with qtcreator and found that the problem came from the lines
"buffer.append(char(0))" of the libs/xml.cpp file.
It works with qt5 but you have to remove them with qt6 without knowing why.

Qt5  parsing  gives   ->   "version"
Qt6 parsing  gives    ->   "version\000"   with  buffer.append(char(0)) .

I will be able to continue my tests and maybe make a pullrequest if all goes well.

Thanks for your help

**Reply by spamatica** [reply]
awesome!
